# Importación de librerias

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.simplefilter('ignore')
import seaborn as sns

# Cargo el dataset tras descargarlo desde https://microdata.worldbank.org/index.php/catalog/394/data-dictionary/F8?file_name=FinStructure_2012_September_Update2x

In [2]:
df = pd.read_csv(r"C:\Users\Victor\Documents\Data_Eda\FinStructure_2012_September_Update2x.csv")

## Examino el dataset

In [3]:
df.head(5)

cn cncode  year      region                 incgr   dbacba  \
0  Afghanistan    AFG  1960  South Asia  Low-income economies   39.184   
1  Afghanistan    AFG  1961  South Asia  Low-income economies    30.73   
2  Afghanistan    AFG  1962  South Asia  Low-income economies   22.677   
3  Afghanistan    AFG  1963  South Asia  Low-income economies  20.1945   
4  Afghanistan    AFG  1964  South Asia  Low-income economies  18.1929   

                   llgdp                 cbagdp                 dbagdp  \
0  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
1  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
2  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
3  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
4  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   

                  ofagdp  ...            stvaltraded             stturnover  \
0  1.79769313486232e+308  ...  1.79769313486232e+308  1.79769313486232e+308   
1  1.79769313486232e+308  ...  1.79769313486232e+308  1.79769313486232e+308   
2  1.79769313486232e+308  ...  1.79769313486232e+308  1.79769313486232e+308   
3  1.79769313486232e+308  ...  1.79769313486232e+308  1.79769313486232e+308   
4  1.79769313486232e+308  ...  1.79769313486232e+308  1.79769313486232e+308   

               listco_pc                 prbond                 pubond  \
0  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
1  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
2  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
3  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
4  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   

                intldebt            intldebtnet                nrbloan  \
0  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
1  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
2  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
3  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   
4  1.79769313486232e+308  1.79769313486232e+308  1.79769313486232e+308   

                  offdep                  remit  
0  1.79769313486232e+308  1.79769313486232e+308  
1  1.79769313486232e+308  1.79769313486232e+308  
2  1.79769313486232e+308  1.79769313486232e+308  
3  1.79769313486232e+308  1.79769313486232e+308  
4  1.79769313486232e+308  1.79769313486232e+308  

[5 rows x 36 columns]

In [4]:
df.shape

(10353, 36)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10353 entries, 0 to 10352
Data columns (total 36 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   cn             10353 non-null  object
 1   cncode         10353 non-null  object
 2   year           10353 non-null  int64 
 3   region         10302 non-null  object
 4   incgr          10353 non-null  object
 5   dbacba         10353 non-null  object
 6   llgdp          10353 non-null  object
 7   cbagdp         10353 non-null  object
 8   dbagdp         10353 non-null  object
 9   ofagdp         10353 non-null  object
 10  pcrdbgdp       10353 non-null  object
 11  pcrdbofgdp     10353 non-null  object
 12  bdgdp          10353 non-null  object
 13  fdgdp          10353 non-null  object
 14  bcbd           10353 non-null  object
 15  ll_usd         10353 non-null  object
 16  overhead       10353 non-null  object
 17  netintmargin   10353 non-null  object
 18  concentration  10353 non-n

# Reviso si hay NaNs

In [6]:
fields = ['cn','cncode','year','region','incgr', 'dbacba','llgdp','cbagdp','dbagdp','ofagdp','pcrdbgdp','pcrdbofgdp','bdgdp',
          'fdgdp','bcbd','ll_usd','overhead','netintmargin','concentration','roa','roe','costinc','zscore','inslife',
          'insnonlife','stmktcap','stvaltraded','stturnover','listco_pc','prbond','pubond','intldebt','intldebtnet','nrbloan',
          'offdep','remit']

In [7]:
for f in fields:
    print(df[f].isna().value_counts())

False    10353
Name: cn, dtype: int64
False    10353
Name: cncode, dtype: int64
False    10353
Name: year, dtype: int64
False    10302
True        51
Name: region, dtype: int64
False    10353
Name: incgr, dtype: int64
False    10353
Name: dbacba, dtype: int64
False    10353
Name: llgdp, dtype: int64
False    10353
Name: cbagdp, dtype: int64
False    10353
Name: dbagdp, dtype: int64
False    10353
Name: ofagdp, dtype: int64
False    10353
Name: pcrdbgdp, dtype: int64
False    10353
Name: pcrdbofgdp, dtype: int64
False    10353
Name: bdgdp, dtype: int64
False    10353
Name: fdgdp, dtype: int64
False    10353
Name: bcbd, dtype: int64
False    10353
Name: ll_usd, dtype: int64
False    10353
Name: overhead, dtype: int64
False    10353
Name: netintmargin, dtype: int64
False    10353
Name: concentration, dtype: int64
False    10353
Name: roa, dtype: int64
False    10353
Name: roe, dtype: int64
False    10353
Name: costinc, dtype: int64
False    10353
Name: zscore, dtype: int64
False    10353


In [8]:
df.region.unique()

array(['South Asia', 'Europe and Central Asia',
       'Middle East and North Africa', 'High-income nonOECD members',
       'Sub-Saharan Africa', 'Latin America & the Caribbean',
       'High-income OECD members', 'East Asia and Pacific', nan],
      dtype=object)

#### Veo que solo hay NaN en region

In [9]:
regionNaN = df.region.isna()==True
df[regionNaN]

cn cncode  year region                          incgr  \
8568  St. Kitts and Nevis    KNA  1960    NaN  Upper-middle-income economies   
8569  St. Kitts and Nevis    KNA  1961    NaN  Upper-middle-income economies   
8570  St. Kitts and Nevis    KNA  1962    NaN  Upper-middle-income economies   
8571  St. Kitts and Nevis    KNA  1963    NaN  Upper-middle-income economies   
8572  St. Kitts and Nevis    KNA  1964    NaN  Upper-middle-income economies   
8573  St. Kitts and Nevis    KNA  1965    NaN  Upper-middle-income economies   
8574  St. Kitts and Nevis    KNA  1966    NaN  Upper-middle-income economies   
8575  St. Kitts and Nevis    KNA  1967    NaN  Upper-middle-income economies   
8576  St. Kitts and Nevis    KNA  1968    NaN  Upper-middle-income economies   
8577  St. Kitts and Nevis    KNA  1969    NaN  Upper-middle-income economies   
8578  St. Kitts and Nevis    KNA  1970    NaN  Upper-middle-income economies   
8579  St. Kitts and Nevis    KNA  1971    NaN  Upper-middle-income economies   
8580  St. Kitts and Nevis    KNA  1972    NaN  Upper-middle-income economies   
8581  St. Kitts and Nevis    KNA  1973    NaN  Upper-middle-income economies   
8582  St. Kitts and Nevis    KNA  1974    NaN  Upper-middle-income economies   
8583  St. Kitts and Nevis    KNA  1975    NaN  Upper-middle-income economies   
8584  St. Kitts and Nevis    KNA  1976    NaN  Upper-middle-income economies   
8585  St. Kitts and Nevis    KNA  1977    NaN  Upper-middle-income economies   
8586  St. Kitts and Nevis    KNA  1978    NaN  Upper-middle-income economies   
8587  St. Kitts and Nevis    KNA  1979    NaN  Upper-middle-income economies   
8588  St. Kitts and Nevis    KNA  1980    NaN  Upper-middle-income economies   
8589  St. Kitts and Nevis    KNA  1981    NaN  Upper-middle-income economies   
8590  St. Kitts and Nevis    KNA  1982    NaN  Upper-middle-income economies   
8591  St. Kitts and Nevis    KNA  1983    NaN  Upper-middle-income economies   
8592  St. Kitts and Nevis    KNA  1984    NaN  Upper-middle-income economies   
8593  St. Kitts and Nevis    KNA  1985    NaN  Upper-middle-income economies   
8594  St. Kitts and Nevis    KNA  1986    NaN  Upper-middle-income economies   
8595  St. Kitts and Nevis    KNA  1987    NaN  Upper-middle-income economies   
8596  St. Kitts and Nevis    KNA  1988    NaN  Upper-middle-income economies   
8597  St. Kitts and Nevis    KNA  1989    NaN  Upper-middle-income economies   
8598  St. Kitts and Nevis    KNA  1990    NaN  Upper-middle-income economies   
8599  St. Kitts and Nevis    KNA  1991    NaN  Upper-middle-income economies   
8600  St. Kitts and Nevis    KNA  1992    NaN  Upper-middle-income economies   
8601  St. Kitts and Nevis    KNA  1993    NaN  Upper-middle-income economies   
8602  St. Kitts and Nevis    KNA  1994    NaN  Upper-middle-income economies   
8603  St. Kitts and Nevis    KNA  1995    NaN  Upper-middle-income economies   
8604  St. Kitts and Nevis    KNA  1996    NaN  Upper-middle-income economies   
8605  St. Kitts and Nevis    KNA  1997    NaN  Upper-middle-income economies   
8606  St. Kitts and Nevis    KNA  1998    NaN  Upper-middle-income economies   
8607  St. Kitts and Nevis    KNA  1999    NaN  Upper-middle-income economies   
8608  St. Kitts and Nevis    KNA  2000    NaN  Upper-middle-income economies   
8609  St. Kitts and Nevis    KNA  2001    NaN  Upper-middle-income economies   
8610  St. Kitts and Nevis    KNA  2002    NaN  Upper-middle-income economies   
8611  St. Kitts and Nevis    KNA  2003    NaN  Upper-middle-income economies   
8612  St. Kitts and Nevis    KNA  2004    NaN  Upper-middle-income economies   
8613  St. Kitts and Nevis    KNA  2005    NaN  Upper-middle-income economies   
8614  St. Kitts and Nevis    KNA  2006    NaN  Upper-middle-income economies   
8615  St. Kitts and Nevis    KNA  2007    NaN  Upper-middle-income economies   
8616  St. Kitts and Nevis    KNA  2008    NaN  Upper-middle-income economies   
8617  St. Kitts and Nev

Todos los NaN estan vinculados a 'St. Kitts and Nevis' (KNA), que esta en Las Antillas

In [10]:
for f in fields:
    print(df[f].describe())

count           10353
unique            203
top       Afghanistan
freq               51
Name: cn, dtype: object
count     10353
unique      203
top         AFG
freq         51
Name: cncode, dtype: object
count    10353.000000
mean      1985.000000
std         14.720312
min       1960.000000
25%       1972.000000
50%       1985.000000
75%       1998.000000
max       2010.000000
Name: year, dtype: float64
count                  10302
unique                     8
top       Sub-Saharan Africa
freq                    2346
Name: region, dtype: object
count                             10353
unique                                5
top       Lower-middle-income economies
freq                               2856
Name: incgr, dtype: object
count                     10353
unique                     6330
top       1.79769313486232e+308
freq                       3918
Name: dbacba, dtype: object
count                     10353
unique                     5663
top       1.79769313486232e+308
freq      

Aunque no hay NaNs, el valor más frecuente en todo el dataset es 1.79769313486232e+308, que no me sirve y me impedirá trabajar.
Examino los valores de las columnas.

## Limpio el Dataset

### Ademas del NaN identificado en la columna region, el valor '1.79769313486232e+308' se repite en el Dataset y equivale a un NaN

Para limpiar el DF, sustituire este valor en cada columna por la media o la mediana (segun la desviación) de los valores disponibles, calculada segun los diferentes grados de ingreso por pais (incgr) y región (region)

In [11]:
df=df.fillna('Latin America & the Caribbean')
df[regionNaN]

cn cncode  year                         region  \
8568  St. Kitts and Nevis    KNA  1960  Latin America & the Caribbean   
8569  St. Kitts and Nevis    KNA  1961  Latin America & the Caribbean   
8570  St. Kitts and Nevis    KNA  1962  Latin America & the Caribbean   
8571  St. Kitts and Nevis    KNA  1963  Latin America & the Caribbean   
8572  St. Kitts and Nevis    KNA  1964  Latin America & the Caribbean   
8573  St. Kitts and Nevis    KNA  1965  Latin America & the Caribbean   
8574  St. Kitts and Nevis    KNA  1966  Latin America & the Caribbean   
8575  St. Kitts and Nevis    KNA  1967  Latin America & the Caribbean   
8576  St. Kitts and Nevis    KNA  1968  Latin America & the Caribbean   
8577  St. Kitts and Nevis    KNA  1969  Latin America & the Caribbean   
8578  St. Kitts and Nevis    KNA  1970  Latin America & the Caribbean   
8579  St. Kitts and Nevis    KNA  1971  Latin America & the Caribbean   
8580  St. Kitts and Nevis    KNA  1972  Latin America & the Caribbean   
8581  St. Kitts and Nevis    KNA  1973  Latin America & the Caribbean   
8582  St. Kitts and Nevis    KNA  1974  Latin America & the Caribbean   
8583  St. Kitts and Nevis    KNA  1975  Latin America & the Caribbean   
8584  St. Kitts and Nevis    KNA  1976  Latin America & the Caribbean   
8585  St. Kitts and Nevis    KNA  1977  Latin America & the Caribbean   
8586  St. Kitts and Nevis    KNA  1978  Latin America & the Caribbean   
8587  St. Kitts and Nevis    KNA  1979  Latin America & the Caribbean   
8588  St. Kitts and Nevis    KNA  1980  Latin America & the Caribbean   
8589  St. Kitts and Nevis    KNA  1981  Latin America & the Caribbean   
8590  St. Kitts and Nevis    KNA  1982  Latin America & the Caribbean   
8591  St. Kitts and Nevis    KNA  1983  Latin America & the Caribbean   
8592  St. Kitts and Nevis    KNA  1984  Latin America & the Caribbean   
8593  St. Kitts and Nevis    KNA  1985  Latin America & the Caribbean   
8594  St. Kitts and Nevis    KNA  1986  Latin America & the Caribbean   
8595  St. Kitts and Nevis    KNA  1987  Latin America & the Caribbean   
8596  St. Kitts and Nevis    KNA  1988  Latin America & the Caribbean   
8597  St. Kitts and Nevis    KNA  1989  Latin America & the Caribbean   
8598  St. Kitts and Nevis    KNA  1990  Latin America & the Caribbean   
8599  St. Kitts and Nevis    KNA  1991  Latin America & the Caribbean   
8600  St. Kitts and Nevis    KNA  1992  Latin America & the Caribbean   
8601  St. Kitts and Nevis    KNA  1993  Latin America & the Caribbean   
8602  St. Kitts and Nevis    KNA  1994  Latin America & the Caribbean   
8603  St. Kitts and Nevis    KNA  1995  Latin America & the Caribbean   
8604  St. Kitts and Nevis    KNA  1996  Latin America & the Caribbean   
8605  St. Kitts and Nevis    KNA  1997  Latin America & the Caribbean   
8606  St. Kitts and Nevis    KNA  1998  Latin America & the Caribbean   
8607  St. Kitts and Nevis    KNA  1999  Latin America & the Caribbean   
8608  St. Kitts and Nevis    KNA  2000  Latin America & the Caribbean   
8609  St. Kitts and Nevis    KNA  2001  Latin America & the Caribbean   
8610  St. Kitts and Nevis    KNA  2002  Latin America & the Caribbean   
8611  St. Kitts and Nevis    KNA  2003  Latin America & the Caribbean   
8612  St. Kitts and Nevis    KNA  2004  Latin America & the Caribbean   
8613  St. Kitts and Nevis    KNA  2005  Latin America & the Caribbean   
8614  St. Kitts and Nevis    KNA  2006  Latin America & the Caribbean   
8615  St. Kitts and Nevis    KNA  2007  Latin America & the Caribbean   
8616  St. Kitts and Nevis    KNA  2008  Latin America & the Caribbean   
8617  St. Kitts and Nevis    KNA  2009  Latin America & the Caribbean   
8618  St. Kitts and Nevis    KNA  2010  Latin America & the Caribbean   

                              incgr                 dbacba  \
8568  Upper-middle-income economies  1.79769313486232e+308   
8569  Upper-middle-income economies  1.79769313486232e+308   
8570  Upper-middle-income economies  1.7

In [12]:
fields = ['dbacba','llgdp','cbagdp','dbagdp','ofagdp','pcrdbgdp','pcrdbofgdp','bdgdp',
          'fdgdp','bcbd','ll_usd','overhead','netintmargin','concentration','roa','roe','costinc','zscore','inslife',
          'insnonlife','stmktcap','stvaltraded','stturnover','listco_pc','prbond','pubond','intldebt','intldebtnet','nrbloan',
          'offdep','remit']
regions = ['High-income OECD members','High-income nonOECD members','Middle East and North Africa',
           'East Asia and Pacific','Latin America & the Caribbean','Europe and Central Asia','South Asia',
           'Sub-Saharan Africa']

incLevel = ['Low-income economies','Upper-middle-income economies','High-income nonOECD members',
            'Lower-middle-income economies','High-income OECD members']

In [13]:
'''
for year in df.year:
    for f in fields:
        for r in regions:
            for l in incLevel:
                noNaN = df[(df.region == r) & (df.incgr == l) & (df.year == year)][f] != '1.79769313486232e+308'
                media = pd.to_numeric(df[(df.region == r) & (df.incgr == l) & (df.year == year)][noNaN][f], downcast='float').mean()
                mediana = pd.to_numeric(df[(df.region == r) & (df.incgr == l) & (df.year == year)][noNaN][f], downcast='float').median()
                print(f,year,r,l,'varianza is',pd.to_numeric(df[(df.region == r) & (df.incgr == l) & (df.year == year)][noNaN][f], downcast='float').var())
                dif = abs(media - mediana)
                if dif >= media/4:
                    print(f,year,r,l,'Uso la mediana por ser mas representativa')
                    df.loc[(df[f] == '1.79769313486232e+308') & (df['region'] == r) & (df.incgr == l) & (df.year == year), f] = mediana
                if dif < media/4:
                    print(f,year,r,l,'Uso la media por ser mas representativa')
                    df.loc[(df[f] == '1.79769313486232e+308') & (df['region'] == r) & (df.incgr == l) & (df.year == year), f] = media
'''
for f in fields:
    for r in regions:
        noNaN = df[(df.region == r)][f] != '1.79769313486232e+308'
        media = pd.to_numeric(df[(df.region == r)][noNaN][f], downcast='float').mean()
        mediana = pd.to_numeric(df[(df.region == r)][noNaN][f], downcast='float').median()
        print(f,r,'varianza is',pd.to_numeric(df[(df.region == r)][noNaN][f], downcast='float').var())
        dif = abs(media - mediana)
        if dif >= media/4:
            print(f,r,'Uso la mediana por ser mas representativa')
            df.loc[(df[f] == '1.79769313486232e+308') & (df['region'] == r), f] = mediana
        if dif < media/4:
            print(f,r,'Uso la media por ser mas representativa')
            df.loc[(df[f] == '1.79769313486232e+308') & (df['region'] == r), f] = media

dbacba High-income OECD members varianza is 116.359215
dbacba High-income OECD members Uso la media por ser mas representativa
dbacba High-income nonOECD members varianza is 131.18874
dbacba High-income nonOECD members Uso la media por ser mas representativa
dbacba Middle East and North Africa varianza is 390.88785
dbacba Middle East and North Africa Uso la media por ser mas representativa
dbacba East Asia and Pacific varianza is 394.62363
dbacba East Asia and Pacific Uso la media por ser mas representativa
dbacba Latin America & the Caribbean varianza is 390.8808
dbacba Latin America & the Caribbean Uso la media por ser mas representativa
dbacba Europe and Central Asia varianza is 312.33356
dbacba Europe and Central Asia Uso la media por ser mas representativa
dbacba South Asia varianza is 556.5338
dbacba South Asia Uso la media por ser mas representativa
dbacba Sub-Saharan Africa varianza is 556.52783
dbacba Sub-Saharan Africa Uso la media por ser mas representativa
llgdp High-income

fdgdp Sub-Saharan Africa varianza is 230.24068
fdgdp Sub-Saharan Africa Uso la media por ser mas representativa
bcbd High-income OECD members varianza is 2144.4702
bcbd High-income OECD members Uso la media por ser mas representativa
bcbd High-income nonOECD members varianza is 1573.8824
bcbd High-income nonOECD members Uso la media por ser mas representativa
bcbd Middle East and North Africa varianza is 1943.4218
bcbd Middle East and North Africa Uso la media por ser mas representativa
bcbd East Asia and Pacific varianza is 8547.474
bcbd East Asia and Pacific Uso la media por ser mas representativa
bcbd Latin America & the Caribbean varianza is 1984.8588
bcbd Latin America & the Caribbean Uso la media por ser mas representativa
bcbd Europe and Central Asia varianza is 9344.278
bcbd Europe and Central Asia Uso la media por ser mas representativa
bcbd South Asia varianza is 2463.7998
bcbd South Asia Uso la media por ser mas representativa
bcbd Sub-Saharan Africa varianza is 4464.052
bcb

inslife Europe and Central Asia varianza is 0.038091492
inslife Europe and Central Asia Uso la media por ser mas representativa
inslife South Asia varianza is 0.9760092
inslife South Asia Uso la mediana por ser mas representativa
inslife Sub-Saharan Africa varianza is 15.540514
inslife Sub-Saharan Africa Uso la mediana por ser mas representativa
insnonlife High-income OECD members varianza is 1.2114506
insnonlife High-income OECD members Uso la media por ser mas representativa
insnonlife High-income nonOECD members varianza is 3.3928323
insnonlife High-income nonOECD members Uso la media por ser mas representativa
insnonlife Middle East and North Africa varianza is 0.39262164
insnonlife Middle East and North Africa Uso la media por ser mas representativa
insnonlife East Asia and Pacific varianza is 0.25135627
insnonlife East Asia and Pacific Uso la media por ser mas representativa
insnonlife Latin America & the Caribbean varianza is 0.5199763
insnonlife Latin America & the Caribbean Us

intldebtnet Latin America & the Caribbean varianza is 1.4956652
intldebtnet Latin America & the Caribbean Uso la mediana por ser mas representativa
intldebtnet Europe and Central Asia varianza is 1.0716879
intldebtnet Europe and Central Asia Uso la mediana por ser mas representativa
intldebtnet South Asia varianza is 0.276806
intldebtnet South Asia Uso la mediana por ser mas representativa
intldebtnet Sub-Saharan Africa varianza is 13.924242
intldebtnet Sub-Saharan Africa Uso la mediana por ser mas representativa
nrbloan High-income OECD members varianza is 36203.51
nrbloan High-income OECD members Uso la mediana por ser mas representativa
nrbloan High-income nonOECD members varianza is 2074789.8
nrbloan High-income nonOECD members Uso la mediana por ser mas representativa
nrbloan Middle East and North Africa varianza is 51.391834
nrbloan Middle East and North Africa Uso la media por ser mas representativa
nrbloan East Asia and Pacific varianza is 3450026.966666393
nrbloan East Asia an

In [14]:
df

cn cncode  year              region                 incgr  \
0      Afghanistan    AFG  1960          South Asia  Low-income economies   
1      Afghanistan    AFG  1961          South Asia  Low-income economies   
2      Afghanistan    AFG  1962          South Asia  Low-income economies   
3      Afghanistan    AFG  1963          South Asia  Low-income economies   
4      Afghanistan    AFG  1964          South Asia  Low-income economies   
...            ...    ...   ...                 ...                   ...   
10348     Zimbabwe    ZWE  2006  Sub-Saharan Africa  Low-income economies   
10349     Zimbabwe    ZWE  2007  Sub-Saharan Africa  Low-income economies   
10350     Zimbabwe    ZWE  2008  Sub-Saharan Africa  Low-income economies   
10351     Zimbabwe    ZWE  2009  Sub-Saharan Africa  Low-income economies   
10352     Zimbabwe    ZWE  2010  Sub-Saharan Africa  Low-income economies   

          dbacba      llgdp    cbagdp     dbagdp   ofagdp  ... stvaltraded  \
0         39.184  36.605648  8.923634  26.232584  1.54947  ...     2.93113   
1          30.73  36.605648  8.923634  26.232584  1.54947  ...     2.93113   
2         22.677  36.605648  8.923634  26.232584  1.54947  ...     2.93113   
3        20.1945  36.605648  8.923634  26.232584  1.54947  ...     2.93113   
4        18.1929  36.605648  8.923634  26.232584  1.54947  ...     2.93113   
...          ...        ...       ...        ...      ...  ...         ...   
10348  69.377403   24.73056   5.38915  18.747852  437.845  ...     18.2396   
10349  69.377403   24.73056   5.38915  18.747852  3.45002  ...     754.028   
10350  69.377403   24.73056   5.38915  18.747852  3.45002  ...    0.560719   
10351  69.377403   24.73056   5.38915  18.747852  3.45002  ...    0.560719   
10352  69.377403   24.73056   5.38915  18.747852  3.45002  ...    0.560719   

      stturnover listco_pc     prbond     pubond  intldebt intldebtnet  \
0      23.544201  0.059603    0.90993  28.129766  1.529482    0.254631   
1      23.544201  0.059603    0.90993  28.129766  1.529482    0.254631   
2      23.544201  0.059603    0.90993  28.129766  1.529482    0.254631   
3      23.544201  0.059603    0.90993  28.129766  1.529482    0.254631   
4      23.544201  0.059603    0.90993  28.129766  1.529482    0.254631   
...          ...       ...        ...        ...       ...         ...   
10348    99.1067  0.063849  14.138269  41.524384   5.50744     0.25927   
10349     201.86  0.065699  14.138269  41.524384   5.50744     0.25927   
10350   2.473905  0.062643  14.138269  41.524384   5.50744     0.25927   
10351   2.473905  0.060927  14.138269  41.524384   5.50744     0.25927   
10352   2.473905  0.060454  14.138269  41.524384   5.50744     0.25927   

        nrbloan    offdep     remit  
0      6.214133   4.39015   2.91073  
1      6.214133   4.39015   2.91073  
2      6.214133   4.39015   2.91073  
3      6.214133   4.39015   2.91073  
4      6.214133   4.39015   2.91073  
...         ...       ...       ...  
10348   15.3934  0.328146  0.949945  
10349    16.459  0.001176  0.949945  
10350   18.1851   45.2698  0.949945  
10351   6.35686   45.2698  0.949945  
10352   3.87908   45.2698  0.949945  

[10353 rows x 36 columns]

In [15]:
for f in fields:
    df[f] = pd.to_numeric(df[f], downcast='float')

ValueError: Unable to parse string "1.79769313486232e+308" at position 102

### Busco el motivo del fallo en el bucle

In [16]:
for f in fields:
    for v in range(len(df[f])-1):
        if df[f][v] == '1.79769313486232e+308':
            print(f,df[f][v])


prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769313486232e+308
prbond 1.79769

In [17]:
df[df['prbond'] == '1.79769313486232e+308']

cn cncode  year                        region  \
102        Algeria    DZA  1960  Middle East and North Africa   
103        Algeria    DZA  1961  Middle East and North Africa   
104        Algeria    DZA  1962  Middle East and North Africa   
105        Algeria    DZA  1963  Middle East and North Africa   
106        Algeria    DZA  1964  Middle East and North Africa   
...            ...    ...   ...                           ...   
10246  Yemen, Rep.    YEM  2006  Middle East and North Africa   
10247  Yemen, Rep.    YEM  2007  Middle East and North Africa   
10248  Yemen, Rep.    YEM  2008  Middle East and North Africa   
10249  Yemen, Rep.    YEM  2009  Middle East and North Africa   
10250  Yemen, Rep.    YEM  2010  Middle East and North Africa   

                               incgr     dbacba      llgdp    cbagdp  \
102    Upper-middle-income economies  75.590569  53.644871  9.260660   
103    Upper-middle-income economies  75.590569  53.644871  9.260660   
104    Upper-middle-income economies  75.590569  53.644871  9.260660   
105    Upper-middle-income economies  75.590569  53.644871  9.260660   
106    Upper-middle-income economies  52.303398  53.644871  9.260660   
...                              ...        ...        ...       ...   
10246  Lower-middle-income economies  99.288101  30.577299  0.056378   
10247  Lower-middle-income economies  99.936401  31.854700  0.044924   
10248  Lower-middle-income economies  86.592598  31.473700  0.984337   
10249  Lower-middle-income economies  74.016403  35.645302  4.145720   
10250  Lower-middle-income economies  68.510902  29.574400  5.763740   

         dbagdp    ofagdp  ...  stvaltraded  stturnover  listco_pc  \
102    37.75526  6.644979  ...     3.023145     13.1943   0.047496   
103    37.75526  6.644979  ...     3.023145     13.1943   0.047496   
104    37.75526  6.644979  ...     3.023145     13.1943   0.047496   
105    37.75526  6.644979  ...     3.023145     13.1943   0.047496   
106    37.75526  6.644979  ...     3.023145     13.1943   0.047496   
...         ...       ...  ...          ...         ...        ...   
10246  11.54560  6.644979  ...     3.023145     13.1943   0.047496   
10247  12.68560  6.644979  ...     3.023145     13.1943   0.047496   
10248  13.04830  6.644979  ...     3.023145     13.1943   0.047496   
10249  15.74240  6.644979  ...     3.023145     13.1943   0.047496   
10250  14.23750  6.644979  ...     3.023145     13.1943   0.047496   

                      prbond    pubond  intldebt  intldebtnet   nrbloan  \
102    1.79769313486232e+308  77.99395   2.88066      1.48343  8.937109   
103    1.79769313486232e+308  77.99395   2.88066      1.48343  8.937109   
104    1.79769313486232e+308  77.99395   2.88066      1.48343  8.937109   
105    1.79769313486232e+308  77.99395   2.88066      1.48343  8.937109   
106    1.79769313486232e+308  77.99395   2.88066      1.48343  8.937109   
...                      ...       ...       ...          ...       ...   
10246  1.79769313486232e+308  77.99395   2.88066      1.48343   1.35733   
10247  1.79769313486232e+308  77.99395   2.88066      1.48343   1.72696   
10248  1.79769313486232e+308  77.99395   2.88066      1.48343   3.74851   
10249  1.79769313486232e+308  77.99395   2.88066      1.48343   4.81097   
10250  1.79769313486232e+308  77.99395   2.88066      1.48343   4.57434   

        offdep    remit  
102    14.0045  4.39879  
103    14.0045  4.39879  
104    14.0045  4.39879  
105    14.0045  4.39879  
106    14.0045  4.39879  
...        ...      ...  
10246  22.3985  6.72164  
10247  24.7335  6.10208  
10248  15.8599  5.24011  
10249  17.0232  4.61598  
10250  28.9161  3.99371  

[663 rows x 36 columns]

In [18]:
df[df['prbond'] == '1.79769313486232e+308']['prbond'].value_counts()

1.79769313486232e+308    663
Name: prbond, dtype: int64

### Falla el bucle en prbond porque se han quedado 663 valores '1.79769313486232e+308' vinculados a 'Middle East and North Africa'	'Lower-middle-income economies'

Limpio la columna de forma individualizada

In [19]:
noNaN = df[df.incgr == 'Lower-middle-income economies']['prbond'] != '1.79769313486232e+308'
media = pd.to_numeric(df[df.incgr == 'Lower-middle-income economies'][noNaN]['prbond'], downcast='float').mean()
mediana = pd.to_numeric(df[df.incgr == 'Lower-middle-income economies'][noNaN]['prbond'], downcast='float').median()
print(media)
print(mediana)
print('varianza is',pd.to_numeric(df[df.incgr == 'Lower-middle-income economies'][noNaN]['prbond'], downcast='float').var())
dif = abs(media - mediana)
print(dif)
'''        
        if dif >= media/4:
            print('Uso la mediana por ser mas representativa')
            df.loc[(df['prbond'] == '1.79769313486232e+308') & (df['region'] == 'Middle East and North Africa') & (df.incgr == 'Lower-middle-income economies'), ['prbond']] = mediana
        if dif < media/4:
            print('Uso la media por ser mas representativa')
            df.loc[(df['prbond'] == '1.79769313486232e+308') & (df['region'] == 'Middle East and North Africa') & (df.incgr == 'Lower-middle-income economies'), ['prbond']] = media
'''
noNaN = df[df.region == 'Middle East and North Africa']['prbond'] != '1.79769313486232e+308'
media = pd.to_numeric(df[df.region == 'Middle East and North Africa'][noNaN]['prbond'], downcast='float').mean()
mediana = pd.to_numeric(df[df.region == 'Middle East and North Africa'][noNaN]['prbond'], downcast='float').median()
print(media)
print(mediana)
print('varianza is',pd.to_numeric(df[df.region == 'Middle East and North Africa'][noNaN]['prbond'], downcast='float').var())
dif = abs(media - mediana)
print(dif)
'''    
    if dif >= media/4:
        print('Uso la mediana por ser mas representativa')
        df.loc[(df['prbond'] == '1.79769313486232e+308') & (df.region == 'Middle East and North Africa'), ['prbond']] = mediana
    if dif < media/4:
        print('Uso la media por ser mas representativa')
        df.loc[(df['prbond'] == '1.79769313486232e+308') & (df.region == 'Middle East and North Africa'), ['prbond']] = media
'''

7.0554175
7.219645
varianza is 26.332304
0.16422749
nan
nan
varianza is nan
nan


"    \n    if dif >= media/4:\n        print('Uso la mediana por ser mas representativa')\n        df.loc[(df['prbond'] == '1.79769313486232e+308') & (df.region == 'Middle East and North Africa'), ['prbond']] = mediana\n    if dif < media/4:\n        print('Uso la media por ser mas representativa')\n        df.loc[(df['prbond'] == '1.79769313486232e+308') & (df.region == 'Middle East and North Africa'), ['prbond']] = media\n"

Por region no consigue calcular la media y la mediana de paises similares, así que lo hare por nivel de ingresos

In [20]:
for l in incLevel:
    noNaN = df[df.incgr == l]['prbond'] != '1.79769313486232e+308'
    media = pd.to_numeric(df[df.incgr == l][noNaN]['prbond'], downcast='float').mean()
    mediana = pd.to_numeric(df[df.incgr == l][noNaN]['prbond'], downcast='float').median()
    print(media)
    print(mediana)
    print('varianza is',pd.to_numeric(df[df.incgr == l][noNaN]['prbond'], downcast='float').var())
    dif = abs(media - mediana)
    print(dif)

    for i in range(len(df.loc[(df['prbond'] == '1.79769313486232e+308') & (df['region'] == 'Middle East and North Africa') & (df.incgr == l), ['prbond']])-1):

        if dif >= media/4:
            print('Uso la mediana por ser mas representativa')
            df.loc[(df['prbond'] == '1.79769313486232e+308') & (df['region'] == 'Middle East and North Africa') & (df.incgr == l), ['prbond']] = mediana
        if dif < media/4:
            print('Uso la media por ser mas representativa')
            df.loc[(df['prbond'] == '1.79769313486232e+308') & (df['region'] == 'Middle East and North Africa') & (df.incgr == l), ['prbond']] = media

11.317533
14.138269
varianza is 25.388742
2.820737
4.7458286
3.88526
varianza is 31.531126
0.8605685
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por se

Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa


Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa


Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
Uso la media por ser mas representativa
39.570637
39.570637
varianza is 399.14438
0.0


# Confirmo que ha limpiado todos los valores erroneos en prbond

In [21]:
df.loc[(df['prbond'] == '1.79769313486232e+308') & (df['region'] == 'Middle East and North Africa')]

Empty DataFrame
Columns: [cn, cncode, year, region, incgr, dbacba, llgdp, cbagdp, dbagdp, ofagdp, pcrdbgdp, pcrdbofgdp, bdgdp, fdgdp, bcbd, ll_usd, overhead, netintmargin, concentration, roa, roe, costinc, zscore, inslife, insnonlife, stmktcap, stvaltraded, stturnover, listco_pc, prbond, pubond, intldebt, intldebtnet, nrbloan, offdep, remit]
Index: []

[0 rows x 36 columns]

In [22]:
df[df['prbond'] == '1.79769313486232e+308']

Empty DataFrame
Columns: [cn, cncode, year, region, incgr, dbacba, llgdp, cbagdp, dbagdp, ofagdp, pcrdbgdp, pcrdbofgdp, bdgdp, fdgdp, bcbd, ll_usd, overhead, netintmargin, concentration, roa, roe, costinc, zscore, inslife, insnonlife, stmktcap, stvaltraded, stturnover, listco_pc, prbond, pubond, intldebt, intldebtnet, nrbloan, offdep, remit]
Index: []

[0 rows x 36 columns]

In [23]:
df['prbond'] = pd.to_numeric(df['prbond'], downcast='float')

# Vuelvo a ejecutar el bucle original para convertir todas las columnas a float

In [24]:
for f in fields:
    df[f] = pd.to_numeric(df[f], downcast='float')

# Confirmo que ya puedo trabajar con el dataset y guardo una copia

In [26]:
df.to_csv(r'C:\Users\Victor\Desktop\EDA_Victor_Bandin\src\data\df_EDA_Victor_Bandin.csv', index=False, header=True)